sbt-scalajs-bundler/src/main/scala/scalajsbundler/BundlerFile.scala

```diff
@@ -0,0 +1,164 @@
+    */
+  case class EntryPoint(application: Application, file: java.io.File)
+      extends Internal
+      with WebpackInput {
+    def project = application.project
+  }
+
+  object EntryPoint {
+    /** Filename of the generated bundle, given its module entry name */
+    def fileName(entry: String): String = s"$entry-entrypoint.js"
+  }
+
+  /**
+    * The package.json file, used for populating the node_modules folder
+    *
+    * @param file The file reference for the package.json
+    */
+  case class PackageJson(file: java.io.File) extends Internal
+
+  /**
+    * A webpack configuration file.
+    *
+    * @param application The [[Application]] this file runs webpack for
+    * @param file The webpack.config.js file reference
+    */
+  case class WebpackConfig(application: Application, file: java.io.File)
+      extends Internal {
+    def project = application.project
+
+    def targetDir: File = file.getParentFile
+
+    def asLibrary: Library =
+      Library(project,
+              file.getParentFile.toPath
+                .resolve(Library.fileName(project))
+                .toFile)
+
+    def asApplicationBundle: ApplicationBundle =
+      ApplicationBundle(project,
+                        file.getParentFile.toPath
+                          .resolve(ApplicationBundle.fileName(project))
+                          .toFile)
+
+  }
+
+  /**
+    * External webpack artifacts -- those that might be served to clients or packaged
+    */
+  sealed abstract class External extends BundlerFile {
+    def project: String
+    def `type`: BundlerFileType
+  }
+
+  /**
+    * The scala.js application itself, aka -fastopt.js or -opt.js
+    *
+    * @param project The application project name
+    * @param file The file containing the application javascript
+    */
+  case class Application(project: String, file: java.io.File)
+      extends External
+      with WebpackInput {
+    override val `type`: BundlerFileType = BundlerFileType.Application
+    def asLoader: Loader =
+      Loader(this,
+             file.getParentFile.toPath
+               .resolve(Loader.fileName(project))
+               .toFile)
+
+    def asEntryPoint: EntryPoint =
+      EntryPoint(this,
+                 file.getParentFile.toPath
+                   .resolve(EntryPoint.fileName(project))
+                   .toFile)
+
+    def asApplicationBundle: ApplicationBundle =
+      ApplicationBundle(project,
+                        file.getParentFile.toPath
+                          .resolve(ApplicationBundle.fileName(project))
+                          .toFile)
+  }
+
+  /**
+    * A webpack library bundle, containing only libraries
+    * @param project The project the library bundle was generated for
+    * @param file The file containing the project libraries
+    */
+  case class Library(project: String, file: java.io.File) extends External {
+    override val `type`: BundlerFileType = BundlerFileType.Library
```

Comment:
I prefer to omit the `override` keyword when a member is just “implemented” (not “overriden”)

Reply:
I'll drop it. I usually include override since it ensures if the underlying interfaces changes the compiler will tell me that I now have "unneeded" implementations.